docs/design/2024-10-11-put-and-verify-transactions-for-external-storages.md

```diff
@@ -0,0 +1,233 @@
+	// This will be called when the write is allowed and about to be performed.
+	// If `Verify()` returns an error, the write will be aborted.
+	// 
+	// With this, you may add extra preconditions of committing in usecases like RWLock.
+	Verify func(ctx VerifyWriteContext) error
+}
+```
+
+After successfully committing such a PutAndVerify transaction, the following invariants are kept:
+
+- The `Target` should be written exactly once.
+- The function `Verify()` returns no error after this transaction committed, if the function `Verify` satisfies:
+  - Once it returns non-error, it will always return non-error as long as exactly one file has the Prefix().
+
+A PutAndVerify txn was committed by:
+
+- Put a intention file with a random suffix to the `Target`.
+- Check if there is another file in the `Target`.
+- If there is, remove our intention file and back off or report error to caller.
+- If there isn't, and `Verify()` returns no error, put the `Content()` to the `Target`, without suffix, then remove the intention file.
+
+Here is the detailed code for committing such a transaction:
+
+```go
+func (w ExclusiveWrite) CommitTo(ctx context.Context, s ExternalStorage) (uuid.UUID, error) {
+	txnID := uuid.New()
+	cx := VerifyWriteContext{
+		Context: ctx,
+		Target:  w.Target,
+		Storage: s,
+		TxnID:   txnID,
+	}
+	intentFileName := cx.IntentFileName() // Should be "{Target}.INTENT.{UUID}"
+	checkConflict := func() error {
+		var err error
+		if w.Verify != nil {
+			err = multierr.Append(err, w.Verify(cx))
+		}
+		return multierr.Append(err, cx.assertOnlyMyIntent())
+	}
+
+	if err := checkConflict(); err != nil {
+		return uuid.UUID{}, errors.Annotate(err, "during initial check")
+	}
+	if err := s.WriteFile(cx, intentFileName, []byte{}); err != nil {
+		return uuid.UUID{}, errors.Annotate(err, "during writing intention file")
+	}
+	defer s.DeleteFile(cx, intentFileName)
+	if err := checkConflict(); err != nil {
+		return uuid.UUID{}, errors.Annotate(err, "during checking whether there are other intentions")
+	}
+
+	return txnID, s.WriteFile(cx, w.Target, w.Content(txnID))
+}
+```
+
+
+
+An example of the txn aborting, when there are two conflicting txns, the name of intent files are simplified for reading:
+
+| Alice's Txn                              | Bob's Txn                                  |
+| ---------------------------------------- | ------------------------------------------ |
+| intentFile := "LOCK_Alice"              | intentFile := "LOCK_Bob"                  |
+|                                          | Verify() → **OK**                          |
+| Verify() → **OK**                        |                                            |
+|                                          | Write("LOCK_Bob", "") → **OK**             |
+| Write("LOCK_Alice", "") → **OK**         |                                            |
+|                                          | Verify() → **Failed! "LOCK_Alice" exists** |
```

Comment:
When contention is high no one can commit? It doesn't scale.

Reply:
Perhaps a randomized back off algorithm can help. Anyway this is like some sort of optimistic lock, hence not pretty effective in the scenario that tons of clients racing for one lock...

Reply:
OK if there will not be too many clients in the use case